### ISSUE TITLE
Is it possible to add attention_mask in decode function?

### ISSUE THREAD

**Issue by xray1111** [NVIDIA/TensorRT-LLM#364]
Thank you guys for the fantastic framework!

But here I have some questions:

If I have batch size of sentences (say 20 sentences, and the char number of sentence is not the same) to be feed to an engine at the same time, and I am not setting 'remove_input_padding' while building the engine, then I have to pad special tokens(like 0) to short sentences to build a valid 2-d tensor before decode. Then:

1. Is there any way to indicate which side (left or right) from where I can pad specific tokens? because the padding side is related with my model's training method.
2. If I pad some token, I need to use a specific attention mask to mask out the padding tokens before inference

I don't see any related input paramter of decode function could do the above 2 things, hope to get your help!

**Comment by byshiue** [comment]
1. TensorRT LLM only supports right side padding now. 
2. TensorRT LLM requires a inputs called input_lengths. TensorRT LLM will generate attention mask by input_lengths for non-attention-plugin case. In attention-plugin case, the plugin can skip the padding part by input_lengths directly.

**Comment by xray1111** [comment]
Thank you for your quick reply!

Do you think it's necessary to add left side padding option? If it is, how can I implement it decently? I may do some work and make a PR.
@byshiue

**Comment by byshiue** [comment]
Is there any reason to ask left padding? It would affect lots of kernels related to attention part. So, we don't have plan to do it now.

**Comment by xray1111** [comment]
> Is there any reason to ask left padding? It would affect lots of kernels related to attention part. So, we don't have plan to do it now.

Yes I think it is necessory, because in training or fine-tuning stage, we normally construct batch of prompts, which usually not the same length, together with the target sentences. And the target sentences are always right after the prompts immediately, that means there is no other spaces or chars between the prompt and the answer. And the model also learns this.

In inference situation, If we append other characters behind prompt, it would affect the performance of the trained model (I have observed this in my test case, and I think it is due to the above training style).

And if we can append padding tokens left side, and made the batch of prompt tensor rightly aligned, the above problem will not take place.

I guess that is why in huggingface tokenizer api, there is a option to set padding side choice
@byshiue

**Comment by xray1111** [comment]
I guess that is why in huggingface tokenizer api, there is a option to set padding side choice
@byshiue

**Comment by byshiue** [comment]
> If we append other characters behind prompt, it would affect the performance of the trained model

That shouldn't happen. Although we add pad after inputs, but these pad does not affect the computation, we skip these parts directly.

**Comment by xray1111** [comment]
> > If we append other characters behind prompt, it would affect the performance of the trained model
> 
> That shouldn't happen. Although we add pad after inputs, but these pad does not affect the computation, we skip these parts directly.

Sure we pad tokens and also mark it as useless by attention_mask, but in the training stage, model never see such kind of sentences which is like 'prompts + padding tokens + answers'. All the training corpus are like 'prompts + answers + padding tokens', and the corresponding attention mask are also consecutive. While in the previous style, the attention mask is not consecutive, it is cut into two parts by padding tokens.

And I believe this kind of difference between training and inference will degrade model's inference performance .
@byshiue

**Comment by byshiue** [comment]
No. No matter what kind of padding we use in training, the real computation logics always  `prompts + answers` without padding. So, it shouldn't affect the results no matter what pattern of padding we use, unless the implementation has bug. Otherwise, the number of padding may also affect the results and that's not reasonable. 

So, we only ensure the computation logic in TRT LLM is same to `prompt + answers` no matter we have right padding or not (remove padding case).

**Comment by xray1111** [comment]
I'm very sorry to take so much time of yours, but I still want to make it clearer to understand, please correct me if I am wrong, thank you very much!

Say I have 3 tokenized prompts:
`
[[32, 782, 110],
[1110, 48, 71, 344, 310]
[31, 9]]
`
before they were sent to the engine, they must be padded as (if right side padding):
`
[[32, 782, 110, 0, 0],
[1110, 48, 71, 344, 310]
[31, 9, 0, 0, 0]]
`
and the attention mask will be:
`
[[1, 1, 1, 0, 0],
[1, 1, 1, 1, 1]
[1, 1, 0, 0, 0]]
`
after decoding, the generated answer will be appended to the padded tensor as (if the max number of tokens is set to 10):
`
[[32, 782, 110, 0, 0, a1, a2, a3, a4, a5],
[1110, 48, 71, 344, 310, b1, b2, b3, b4, b5]
[31, 9, 0, 0, 0, c1, c2, c3, c4, c5]]
`
So, we can see that the 1st and 3rd generated answers are not after the promps immediately, but is seperated by padding tokens.
@byshiue

**Comment by xray1111** [comment]
Here is another explaination on why we should use left padding when generating text with a decoder-only language model, would you please take a look at it? Thank you!

https://github.com/https://github.com/huggingface/transformers/issues/18388#issuecomment-1204369688

@byshiue

**Comment by byshiue** [comment]
Although the ids under right padding may be like 
```
[
[32, 782, 110, 0, 0, a1, a2, a3, a4, a5], 
[1110, 48, 71, 344, 310, b1, b2, b3, b4, b5],
[31, 9, 0, 0, 0, c1, c2, c3, c4, c5]
]
```

but the computation workflow/logic/order should fully same to 
```
[[32, 782, 110, a1, a2, a3, a4, a5], [1110, 48, 71, 344, 310, b1, b2, b3, b4, b5] [31, 9, c1, c2, c3, c4, c5]]
```

In fact, when the `remove_input_padding` is enabled, the inputs are concanenated as 
```
[[32, 782, 110, 1110, 48, 71, 344, 310, 31, 9]]
```

Also, the outputs are also puts as 
```
[
[32, 782, 110, a1, a2, a3, a4, a5, 0, 0],
[1110, 48, 71, 344, 310, b1, b2, b3, b4, b5]
[31, 9, c1, c2, c3, c4, c5, 0, 0, 0]
]
```
no matter we use `remove_input_padding` or not. For first sentence, the results would be like
```
step 0: [32, 782, 110, 0, 0, 0, 0, 0, 0, 0],
step 1: [32, 782, 110, a1, 0, 0, 0, 0, 0, 0],
step 1: [32, 782, 110, a1, a2, 0, 0, 0, 0, 0],
...
```

**Comment by xray1111** [comment]
> Although the ids under right padding may be like
> 
> ```
> [
> [32, 782, 110, 0, 0, a1, a2, a3, a4, a5], 
> [1110, 48, 71, 344, 310, b1, b2, b3, b4, b5],
> [31, 9, 0, 0, 0, c1, c2, c3, c4, c5]
> ]
> ```
> 
> but the computation workflow/logic/order should fully same to
> 
> ```
> [[32, 782, 110, a1, a2, a3, a4, a5], [1110, 48, 71, 344, 310, b1, b2, b3, b4, b5] [31, 9, c1, c2, c3, c4, c5]]
> ```
> 
> In fact, when the `remove_input_padding` is enabled, the inputs are concanenated as
> 
> ```
> [[32, 782, 110, 1110, 48, 71, 344, 310, 31, 9]]
> ```
> 
> Also, the outputs are also puts as
> 
> ```
> [
> [32, 782, 110, a1, a2, a3, a4, a5, 0, 0],
> [1110, 48, 71, 344, 310, b1, b2, b3, b4, b5]
> [31, 9, c1, c2, c3, c4, c5, 0, 0, 0]
> ]
> ```
> 
> no matter we use `remove_input_padding` or not. For first sentence, the results would be like
> 
> ```
> step 0: [32, 782, 110, 0, 0, 0, 0, 0, 0, 0],
> step 1: [32, 782, 110, a1, 0, 0, 0, 0, 0, 0],
> step 1: [32, 782, 110, a1, a2, 0, 0, 0, 0, 0],
> ...
> ```

Thank you for your kindly reply, I'm trying to look into the code and understand your point.

**Comment by byshiue** [comment]
Close this bug because the question is answered. Feel free to reopen it if needed.
